service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -0,0 +1,162 @@
+    private static final String PATIENT_SEARCH_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:search:patient-1";
+    private static final String DOCUMENT_SEARCH_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:search:documentreference-1";
+    private static final String BINARY_READ_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:read:binary-1";
+    private static final String SSP_INTERACTION_ID = "Ssp-InteractionID";
+    private static final String SLASH = "/";
+    private static final String FHIR_SEPARATOR = "fhir/";
+    private static final String DOCUMENTS_SEPARATOR = "documents/";
+    private static final String DOUBLE_SEPARATOR = DOCUMENTS_SEPARATOR + FHIR_SEPARATOR;
+
+    private final SdsClient sdsClient;
+
+    private Map<String, Function<String, Optional<SdsClient.SdsResponseData>>> sdsRequestFunctions;
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (StringUtils.isBlank(System.getProperty(GPC_URL_ENVIRONMENT_VARIABLE))) {
+            ServerHttpRequest serverHttpRequest = exchange.getRequest();
+            extractInteractionId(serverHttpRequest.getHeaders())
+                .ifPresent(id -> proceedSdsLookup(serverHttpRequest, exchange, id));
+        }
+
+        return chain.filter(exchange);
+    }
+
+    @Override
+    public int getOrder() {
+        return RouteToRequestUrlFilter.ROUTE_TO_URL_FILTER_ORDER + 1;
+    }
+
+    @PostConstruct
+    @SuppressWarnings("unused")
+    public void initializeSdsRequestFunctions() {
+        sdsRequestFunctions = Map.of(
+            STRUCTURED_ID, sdsClient::callForGetStructuredRecord,
+            PATIENT_SEARCH_ID, sdsClient::callForPatientSearchAccessDocument,
+            DOCUMENT_SEARCH_ID, sdsClient::callForSearchForDocumentRecord,
+            BINARY_READ_ID, sdsClient::callForRetrieveDocumentRecord);
+    }
+
+    private void proceedSdsLookup(ServerHttpRequest serverHttpRequest,
+            ServerWebExchange exchange,
+            String integrationId) {
+        Optional<SdsClient.SdsResponseData> response
+            = performRequestAccordingToInteractionId(integrationId, serverHttpRequest);
+
+        if (response.isPresent()) {
+            String address = response.get()
+                .getAddress();
+            prepareLookupUri(address, serverHttpRequest.getPath()).ifPresent(uri -> exchange.getAttributes()
+                .put(ServerWebExchangeUtils.GATEWAY_REQUEST_URL_ATTR, uri));
+        } else {
+            LOGGER.error("SDS filter request was unsuccessful.");
+        }
+    }
+
+    private Optional<SdsClient.SdsResponseData> performRequestAccordingToInteractionId(String interactionId,
+            ServerHttpRequest serverHttpRequest) {
+        if (isRequestFunctionAvailable(interactionId)) {
+            Optional<String> organisation = extractOrganisation(serverHttpRequest
```

Comment:
Don't need all of these Optional and isPresent. Just throw an exception as soon as we know the ODS code is missing. Its an unrecoverable error.